File: lib/indicators.js
'use strict';

// ============================================================================
// indicators.js — deterministic technical indicators computed from OHLCV data.
// All functions take arrays of numbers and return either a full series or the
// latest value(s). No guessing, no image parsing — pure math on real candles.
// ============================================================================

function sma(values, period) {
  if (!values || values.length < period) return null;
  const out = new Array(values.length).fill(null);
  let sum = 0;
  for (let i = 0; i < values.length; i++) {
    sum += values[i];
    if (i >= period) sum -= values[i - period];
    if (i >= period - 1) out[i] = sum / period;
  }
  return out;
}

function ema(values, period) {
  if (!values || values.length < period) return null;
  const out = new Array(values.length).fill(null);
  const k = 2 / (period + 1);
  // Seed with SMA of the first `period` values.
  let seed = 0;
  for (let i = 0; i < period; i++) seed += values[i];
  seed /= period;
  out[period - 1] = seed;
  for (let i = period; i < values.length; i++) {
    out[i] = values[i] * k + out[i - 1] * (1 - k);
  }
  return out;
}

// Wilder's RSI.
function rsi(closes, period = 14) {
  if (!closes || closes.length < period + 1) return null;
  const out = new Array(closes.length).fill(null);
  let gain = 0;
  let loss = 0;
  for (let i = 1; i <= period; i++) {
    const diff = closes[i] - closes[i - 1];
    if (diff >= 0) gain += diff;
    else loss -= diff;
  }
  let avgGain = gain / period;
  let avgLoss = loss / period;
  out[period] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  for (let i = period + 1; i < closes.length; i++) {
    const diff = closes[i] - closes[i - 1];
    const g = diff > 0 ? diff : 0;
    const l = diff < 0 ? -diff : 0;
    avgGain = (avgGain * (period - 1) + g) / period;
    avgLoss = (avgLoss * (period - 1) + l) / period;
    out[i] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  }
  return out;
}

// MACD: returns { macd, signal, histogram } as aligned series.
function macd(closes, fast = 12, slow = 26, signalPeriod = 9) {
  if (!closes || closes.length < slow + signalPeriod) return null;
  const emaFast = ema(closes, fast);
  const emaSlow = ema(closes, slow);
  const macdLine = closes.map((_, i) =>
    emaFast[i] != null && emaSlow[i] != null ? emaFast[i] - emaSlow[i] : null
  );
  const macdValues = macdLine.filter((v) => v != null);
  const signalRaw = ema(macdValues, signalPeriod);
  // Re-align signal to the full length.
  const firstIdx = macdLine.findIndex((v) => v != null);
  const signalLine = new Array(closes.length).fill(null);
  if (signalRaw) {
    for (let i = 0; i < signalRaw.length; i++) {
      if (signalRaw[i] != null) signalLine[firstIdx + i] = signalRaw[i];
    }
  }
  const histogram = closes.map((_, i) =>
    macdLine[i] != null && signalLine[i] != null ? macdLine[i] - signalLine[i] : null
  );
  return { macd: macdLine, signal: signalLine, histogram };
}

// Bollinger Bands: returns { upper, mid, lower, bandwidth } series.
function bollinger(closes, period = 20, mult = 2) {
  if (!closes || closes.length < period) return null;
  const mid = sma(closes, period);
  const upper = new Array(closes.length).fill(null);
  const lower = new Array(closes.length).fill(null);
  const bandwidth = new Array(closes.length).fill(null);
  for (let i = period - 1; i < closes.length; i++) {
    let sumSq = 0;
    const m = mid[i];
    for (let j = i - period + 1; j <= i; j++) sumSq += (closes[j] - m) ** 2;
    const sd = Math.sqrt(sumSq / period);
    upper[i] = m + mult * sd;
    lower[i] = m - mult * sd;
    bandwidth[i] = m !== 0 ? (upper[i] - lower[i]) / m : 0;
  }
  return { upper, mid, lower, bandwidth };
}

// Average True Range (Wilder).
function atr(highs, lows, closes, period = 14) {
  if (!closes || closes.length < period + 1) return null;
  const tr = new Array(closes.length).fill(null);
  tr[0] = highs[0] - lows[0];
  for (let i = 1; i < closes.length; i++) {
    tr[i] = Math.max(
      highs[i] - lows[i],
      Math.abs(highs[i] - closes[i - 1]),
      Math.abs(lows[i] - closes[i - 1])
    );
  }
  const out = new Array(closes.length).fill(null);
  let sum = 0;
  for (let i = 1; i <= period; i++) sum += tr[i];
  out[period] = sum / period;
  for (let i = period + 1; i < closes.length; i++) {
    out[i] = (out[i - 1] * (period - 1) + tr[i]) / period;
  }
  return out;
}

// Rolling average of volume (excludes the current forming candle when asked).
function volumeAverage(volumes, period = 20) {
  return sma(volumes, period);
}

// Rolling VWAP (Volume-Weighted Average Price) over the last `period` bars.
// The intraday "fair value" anchor scalpers watch: price above a rising VWAP is
// a bullish bias, below a falling VWAP is bearish.
function vwap(candles, period = 96) {
  const out = new Array(candles.length).fill(null);
  for (let i = 0; i < candles.length; i++) {
    const start = Math.max(0, i - period + 1);
    let pv = 0;
    let vol = 0;
    for (let j = start; j <= i; j++) {
      const tp = (candles[j].high + candles[j].low + candles[j].close) / 3;
      pv += tp * candles[j].volume;
      vol += candles[j].volume;
    }
    out[i] = vol > 0 ? pv / vol : null;
  }
  return out;
}

// Convenience: last non-null value of a series.
function last(series) {
  if (!series) return null;
  for (let i = series.length - 1; i >= 0; i--) {
    if (series[i] != null) return series[i];
  }
  return null;
}

module.exports = {
  sma,
  ema,
  rsi,
  macd,
  bollinger,
  atr,
  volumeAverage,
  vwap,
  last,
};

File: android/app/src/main/java/ro/signalpilot/android/Indicators.kt
package ro.signalpilot.android

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.sqrt

object Indicators {
    fun sma(values: List<Double>, period: Int): List<Double?>? {
        if (values.size < period) return null
        val out = MutableList<Double?>(values.size) { null }
        var sum = 0.0
        for (i in values.indices) {
            sum += values[i]
            if (i >= period) sum -= values[i - period]
            if (i >= period - 1) out[i] = sum / period
        }
        return out
    }

    fun ema(values: List<Double>, period: Int): List<Double?>? {
        if (values.size < period) return null
        val out = MutableList<Double?>(values.size) { null }
        val k = 2.0 / (period + 1)
        var seed = values.take(period).average()
        out[period - 1] = seed
        for (i in period until values.size) {
            seed = values[i] * k + seed * (1 - k)
            out[i] = seed
        }
        return out
    }

    fun rsi(closes: List<Double>, period: Int = 14): List<Double?>? {
        if (closes.size < period + 1) return null
        val out = MutableList<Double?>(closes.size) { null }
        var gain = 0.0
        var loss = 0.0
        for (i in 1..period) {
            val diff = closes[i] - closes[i - 1]
            if (diff >= 0) gain += diff else loss -= diff
        }
        var avgGain = gain / period
        var avgLoss = loss / period
        out[period] = if (avgLoss == 0.0) 100.0 else 100 - 100 / (1 + avgGain / avgLoss)
        for (i in period + 1 until closes.size) {
            val diff = closes[i] - closes[i - 1]
            val g = if (diff > 0) diff else 0.0
            val l = if (diff < 0) -diff else 0.0
            avgGain = (avgGain * (period - 1) + g) / period
            avgLoss = (avgLoss * (period - 1) + l) / period
            out[i] = if (avgLoss == 0.0) 100.0 else 100 - 100 / (1 + avgGain / avgLoss)
        }
        return out
    }

    data class Macd(val line: List<Double?>, val signal: List<Double?>, val histogram: List<Double?>)

    fun macd(closes: List<Double>, fast: Int = 12, slow: Int = 26, signalPeriod: Int = 9): Macd? {
        if (closes.size < slow + signalPeriod) return null
        val fastLine = ema(closes, fast) ?: return null
        val slowLine = ema(closes, slow) ?: return null
        val line = closes.indices.map { i ->
            if (fastLine[i] != null && slowLine[i] != null) fastLine[i]!! - slowLine[i]!! else null
        }
        val first = line.indexOfFirst { it != null }
        val raw = line.filterNotNull()
        val rawSignal = ema(raw, signalPeriod)
        val signal = MutableList<Double?>(closes.size) { null }
        if (rawSignal != null) rawSignal.forEachIndexed { i, value -> if (value != null) signal[first + i] = value }
        val histogram = closes.indices.map { i ->
            if (line[i] != null && signal[i] != null) line[i]!! - signal[i]!! else null
        }
        return Macd(line, signal, histogram)
    }

    data class Bollinger(
        val upper: List<Double?>,
        val mid: List<Double?>,
        val lower: List<Double?>,
        val bandwidth: List<Double?>,
    )

    fun bollinger(closes: List<Double>, period: Int = 20, mult: Double = 2.0): Bollinger? {
        val mid = sma(closes, period) ?: return null
        val upper = MutableList<Double?>(closes.size) { null }
        val lower = MutableList<Double?>(closes.size) { null }
        val bandwidth = MutableList<Double?>(closes.size) { null }
        for (i in period - 1 until closes.size) {
            val mean = mid[i] ?: continue
            var sumSq = 0.0
            for (j in i - period + 1..i) sumSq += (closes[j] - mean).pow(2)
            val sd = sqrt(sumSq / period)
            upper[i] = mean + mult * sd
            lower[i] = mean - mult * sd
            bandwidth[i] = if (mean != 0.0) (upper[i]!! - lower[i]!!) / mean else 0.0
        }
        return Bollinger(upper, mid, lower, bandwidth)
    }

    fun atr(candles: List<Candle>, period: Int = 14): List<Double?>? {
        if (candles.size < period + 1) return null
        val tr = MutableList(candles.size) { 0.0 }
        tr[0] = candles[0].high - candles[0].low
        for (i in 1 until candles.size) {
            tr[i] = max(candles[i].high - candles[i].low,
                max(abs(candles[i].high - candles[i - 1].close), abs(candles[i].low - candles[i - 1].close)))
        }
        val out = MutableList<Double?>(candles.size) { null }
        out[period] = (1..period).sumOf { tr[it] } / period
        for (i in period + 1 until candles.size) out[i] = (out[i - 1]!! * (period - 1) + tr[i]) / period
        return out
    }

    fun vwap(candles: List<Candle>, period: Int = 96): List<Double?> {
        return candles.indices.map { i ->
            val start = max(0, i - period + 1)
            var pv = 0.0
            var volume = 0.0
            for (j in start..i) {
                val typical = (candles[j].high + candles[j].low + candles[j].close) / 3
                pv += typical * candles[j].volume
                volume += candles[j].volume
            }
            if (volume > 0) pv / volume else null
        }
    }

    fun last(series: List<Double?>?): Double? = series?.asReversed()?.firstOrNull { it != null }
}
